### ISSUE TITLE
haskell-stack-ghc checker poor performance on non stack projects

### ISSUE THREAD

**Issue by turboMaCk** [flycheck/flycheck#1824]
This issue was originally reported in haskell-flycheck however it seems to be related to core fly-check codebase since uninstalling `flycheck-haskell` does not help. This is the original issue https://github.com/flycheck/flycheck-haskell/issues/98

TL;DR On system with `stack` flycheck is using `haskell-stack-ghc` checker even on project which don't use stack which leads to severe performance issues. 

## Checklist
- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description
On a system with `stack` installed there is a significant performance decrease in flycheck performance on projects which are **not** using stack.

## Steps to reproduce
1. install [`haskell-stack`](https://docs.haskellstack.org/en/stable/README/)
2. open project which is **not** using stack (example: https://github.com/turboMaCk/planning-game)
3. Open Haskell file & observe the blocking code on main thread blocking whole editor UI during typing.

- [x] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior
Flycheck should detect that project is not using stack and use `haskell-ghc` backed or it should be possible to disable backend manually using `dir-locals.el`

## Screenshots
---

## System configuration
```
Syntax checkers for buffer AutoIncrement.hs in haskell-mode:

First checker to run:

  haskell-stack-ghc (explicitly selected)
    - may enable:    yes
    - executable:    Found at /run/current-system/sw/bin/stack
    - next checkers: haskell-hlint

Checkers that may run as part of the first checker's chain:

  haskell-hlint
    - may enable:         yes
    - executable:         Found at /nix/store/3baygn2r6s472vhcgm48l7qk8x684hlw-hlint-3.1.6/bin/hlint
    - configuration file: Not found

Checkers that could run if selected:

  haskell-ghc  select
    - may enable:    yes
    - executable:    Found at /nix/store/pw1d4jmlcidp7src5igr3g6zbwzbglhv-ghc-8.8.4-with-packages/bin/ghc
    - next checkers: haskell-hlint


The following syntax checkers are not registered:

  - clojure-cider-eastwood
  - clojure-cider-kibit
  - clojure-cider-typed

Try adding these syntax checkers to `flycheck-checkers'. Flycheck Mode
is enabled. Use M-x universal-argument C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot (package: 20190906.813)
Emacs version:    26.3
System:           x86_64-pc-linux-gnu
Window system:    x
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes

- original issue in flycheck-haskell: https://github.com/flycheck/flycheck-haskell/issues/98
- I've tried to manually blacklist tstack backend in config (`  (add-to-list 'flycheck-disabled-checkers 'haskell-stack-ghc)
`) )which did not help
- @mx-psi is also experiencing this problem
- both me and @mx-psi are running emacs as daemon (at least in my case system.d service)

**Comment by fmdkdd** [comment]
> I've tried to manually blacklist tstack backend in config ( (add-to-list 'flycheck-disabled-checkers 'haskell-stack-ghc)) )which did not help

That should be your solution right there.  I'm curious as to why this doesn't work for you.  Did you try `M-x flycheck-disable-checker` interactively in the offending buffer?

**Comment by turboMaCk** [comment]
Running `M-x flycheck-disable-checker` in buffer does help. Also using `M-x flycheck-verify-setup` to switch checker helps. However for some reason line `(add-to-list 'flycheck-disabled-checkers 'haskell-stack-ghc)` within my config does not. I can try to call it at some other time if that might help?

**Comment by kabuhr** [comment]
Note that the cause of the underlying performance problem is that `flycheck-haskell--find-stack-default-directory` checks for a parent directory containing a `stack.yaml` file and -- if that fails -- invokes `stack --no-install-ghc path --project-root`.  Since this invocation of `stack` doesn't go through the asynchronous code path for invoking the compiler to actually check code, Emacs will freeze for several hundred milliseconds waiting for `stack` to return the path of the global project, which is a problem if `flycheck` is configured to flycheck on every newline.

**Comment by HDButterbrot** [comment]
For a solution that is a bit more clean than just creating an empty`stack.yaml`, one may just create a symlink called `stack.yaml` to the actual `stack.yaml`file in folder returned by `stack --no-install-ghc path --project-root`.

**Comment by bbatsov** [comment]
Closing: `haskell-stack-ghc` performance on non-stack projects is a `flycheck-haskell` issue. Users can disable the checker per-project.
